source/common/http/http2/codec_impl.cc

```diff
@@ -401,8 +401,15 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
       // if local is not complete.
       if (!stream->deferred_reset_.valid()) {
         if (!stream->waiting_for_non_informational_headers_) {
-          ASSERT(stream->remote_end_stream_);
-          stream->decoder_->decodeTrailers(std::move(stream->headers_));
+          if (!stream->remote_end_stream_) {
+            // This indicates we have received more headers frames than Envoy
+            // supports. Even if this is valid HTTP (something like 103 early hints) fail here
+            // rather than trying to push unexpected headers through the Envoy pipeline as that
+            // will likely result in Envoy crashing.
+            throw CodecProtocolException("Unexpected 'trailers' with no end stream.");
```

Comment:
WDYT about resetting the stream vs. throwing a codec protocol exception? The latter will tear down the entire connection? I think it's probably not a big deal either way but throwing it out there. Whatever we do I think we should add a stat for this case similar to the headers_cb_no_stream_ stat so we have something to look at just in case.

Reply:
I think it would be cleaner, but with the defered reset logic I'd be uncomfortable doing that without a regression test, which I think require a bunch more plumbing to allow the codec client to bypass some of our asserts.  I think I'll add a comment explaining that killing the connection, while suboptimal, is better than a crash and safer than a reset.

Reply:
Oh and added a stat but I'm totally failing to find where the h2 codec stats are documented.  Pointers appreciated.